sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/FlatMapGroupsWithState_StateManager.scala

```diff
@@ -0,0 +1,143 @@
+    if (shouldStoreTimestamp) schema.add("timeoutTimestamp", LongType) else schema
+  }
+
+  def getState(store: StateStore, keyRow: UnsafeRow): FlatMapGroupsWithState_StateData = {
+    val stateRow = store.get(keyRow)
+    stateDataForGets.withNew(
+      keyRow, stateRow, getStateObj(stateRow), getTimestamp(stateRow))
+  }
+
+  def putState(store: StateStore, keyRow: UnsafeRow, state: Any, timestamp: Long): Unit = {
+    val stateRow = getStateRow(state)
+    setTimestamp(stateRow, timestamp)
+    store.put(keyRow, stateRow)
+  }
+
+  def removeState(store: StateStore, keyRow: UnsafeRow): Unit = {
+    store.remove(keyRow)
+  }
+
+  def getAllState(store: StateStore): Iterator[FlatMapGroupsWithState_StateData] = {
+    val stateDataForGetAllState = FlatMapGroupsWithState_StateData()
+    store.getRange(None, None).map { pair =>
+      stateDataForGetAllState.withNew(
+        pair.key, pair.value, getStateObjFromRow(pair.value), getTimestamp(pair.value))
+    }
+  }
+
+  private val stateAttributes: Seq[Attribute] = stateSchema.toAttributes
+
+  // Get the serializer for the state, taking into account whether we need to save timestamps
+  private val stateSerializer = {
+    val nestedStateExpr = CreateNamedStruct(
+      stateEncoder.namedExpressions.flatMap(e => Seq(Literal(e.name), e)))
+    if (shouldStoreTimestamp) {
+      Seq(nestedStateExpr, Literal(GroupStateImpl.NO_TIMESTAMP))
+    } else {
+      Seq(nestedStateExpr)
+    }
+  }
+
+  // Get the deserializer for the state. Note that this must be done in the driver, as
+  // resolving and binding of deserializer expressions to the encoded type can be safely done
+  // only in the driver.
+  private val stateDeserializer = {
+    val boundRefToNestedState = BoundReference(nestedStateOrdinal, stateEncoder.schema, true)
+    val deser = stateEncoder.resolveAndBind().deserializer.transformUp {
+      case BoundReference(ordinal, _, _) => GetStructField(boundRefToNestedState, ordinal)
+    }
+    CaseWhen(Seq(IsNull(boundRefToNestedState) -> Literal(null)), elseValue = deser).toCodegen()
+  }
+
+  private lazy val nestedStateOrdinal = 0
+  private lazy val timeoutTimestampOrdinal = 1
+
+  // Converters for translating state between rows and Java objects
+  private lazy val getStateObjFromRow = ObjectOperator.deserializeRowToObject(
+    stateDeserializer, stateAttributes)
+  private lazy val getStateRowFromObj = ObjectOperator.serializeObjectToRow(stateSerializer)
+
+  private lazy val stateDataForGets = FlatMapGroupsWithState_StateData()
+
+  /** Returns the state as Java object if defined */
+  private def getStateObj(stateRow: UnsafeRow): Any = {
+    if (stateRow == null) null
+    // else if (stateRow.isNullAt(nestedStateOrdinal)) null
+    else getStateObjFromRow(stateRow)
+  }
+
+  /** Returns the row for an updated state */
+  private def getStateRow(obj: Any): UnsafeRow = {
+    val row = getStateRowFromObj(obj)
```

Comment:
This part needs work. The current serializer expressions does not set the top level `groupState` column (pointing to nested state columns) in the unsaferow to null. So i am explicitly setting it to null. However, the saved nested columns are still set to a value. Its possible that nested unsaferow can be removed, thus saving space.